lib/actions/field-trip.js

```diff
@@ -249,17 +406,355 @@ function planInbound (request) {
+ *  size assigned to each itinerary.
+ * @return {Object} result.remainingGroupSize The remaining about of people in
+ *  the field trip group that have yet to be assigned to an itinerary.
+ * @return {Object} result.tripsToBanInSubsequentSearches An array of strings
+ *  representing tripIds that should be added to the list of tripIds to ban when
+ *  making requests for additional itineraries from OTP.
+ */
+function checkValidityAndCapacity (state, request) {
+  const fieldTripModuleConfig = getModuleConfig(state, Modules.FIELD_TRIP)
+  const minimumAllowableRemainingCapacity =
+    fieldTripModuleConfig?.minimumAllowableRemainingCapacity || 10
+  const { travelDateTripsInUse, tripHashLookup } = state.callTaker.fieldTrip
+  const activeSearch = getActiveSearch(state)
+
+  // initialize the remaining group size to be the total group size
+  let remainingGroupSize = getGroupSize(request)
+  const assignedItinerariesByResponse = {}
+  const tripsToBanInSubsequentSearches = []
+
+  // iterate through responses as we need to keep track of the response that
+  // each itinerary came from so that they can be updated
+  activeSearch.response.forEach((response, responseIdx) => {
+    if (!response.plan) {
+      // An error might have occurred!
+      return
+    }
+
+    // iterate through itineraries to check validity and assign field trip
+    // groups
+    response.plan.itineraries.forEach((itinerary, idx) => {
+      let itineraryCapacity = Number.POSITIVE_INFINITY
+
+      // check each individual trip to see if there aren't any conflicts
+      itinerary.legs.forEach(leg => {
+        // non-transit legs have infinite capacity
+        if (!isTransit(leg.mode)) {
+          return
+        }
+        const { tripId } = leg
+
+        // this variable is used to track how many other field trips are using a
+        // particular trip
+        let capacityInUse = 0
+
+        // iterate over trips that are already being used by other field trips
+        travelDateTripsInUse.forEach(tripInUse => {
+          // check if the trip is being used by another field trip
+          let sameVehicleTrip = false
+          if (tripId in tripHashLookup && tripInUse.tripHash) {
+            // use the trip hashes if available
+            sameVehicleTrip = (tripHashLookup[tripId] === tripInUse.tripHash)
+          } else {
+            // as fallback, compare the tripId strings
+            sameVehicleTrip = (tripId === tripInUse.tripId)
+          }
+          // not used by another vehicle, so not used by this other field trip
+          if (!sameVehicleTrip) return
+
+          // check if the stop ranges overlap. It is OK if one field trip begins
+          // where the other ends.
+          if (
+            leg.from.stopIndex >= tripInUse.toStopIndex ||
+              leg.to.stopIndex <= tripInUse.fromStopIndex
+          ) {
+            // legs do not overlap, so not used by this other field trip
```

Comment:
Anywhere we have a return statement in this forEach, I think the comment should be changed from: 
```
... so not used by this other field trip
// to
... so this trip/vehicle is free to use
```

Reply:
Refactored in a448093